components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -112,25 +114,55 @@ pub struct LocaleCanonicalizer {
 }
 
 #[inline]
-fn uts35_rule_matches(source: &Locale, ruletype: &LanguageIdentifier) -> bool {
-    (ruletype.language.is_empty() || ruletype.language == source.id.language)
-        && (ruletype.script.is_none() || ruletype.script == source.id.script)
-        && (ruletype.region.is_none() || ruletype.region == source.id.region)
-        && (ruletype.variants.is_empty()
-            || ruletype
-                .variants
-                .iter()
-                .all(|v| source.id.variants.contains(v)))
+fn uts35_rule_matches<I, V, L>(
+    source: &Locale,
+    language: L,
+    script: Option<Script>,
+    region: Option<Region>,
+    variants: I,
+) -> bool
+where
+    I: Iterator<Item = V>,
+    Variant: PartialOrd<V>,
+    Language: PartialEq<L>,
+{
+    (Language::UND == language || source.id.language == language)
+        && (script.is_none() || script == source.id.script)
+        && (region.is_none() || region == source.id.region)
+        && {
+            // Checks if variants are a subset of source variants.
+            // As both iterators are sorted, this can be done linearly.
+            let source_variants = source.id.variants.iter();
+            'outer: for it in variants {
+                for cand in source_variants.as_ref() {
```

Comment:
Minor Issue: I think `source_variants` is an instance of `core::slice::Iter`, so `.as_ref()` converts it back into a `&[Variant]`, and then `for .. in` calls `.into_iter()` on that, so you are never actually advancing the `source_variants` iterator.

Reply:
Good catch, the clippy suggestion I applied was actually for `.by_ref()`